File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/usecases/BrowseCouponsVerification.java
package com.mycompany.myshop.testkit.dsl.core.usecase.usecases;

import com.mycompany.myshop.testkit.dsl.core.shared.ResponseVerification;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;
import com.mycompany.myshop.testkit.common.Converter;
import com.mycompany.myshop.testkit.driver.port.dtos.BrowseCouponsResponse;

import static org.assertj.core.api.Assertions.assertThat;

public class BrowseCouponsVerification extends ResponseVerification<BrowseCouponsResponse> {
    public BrowseCouponsVerification(BrowseCouponsResponse response, UseCaseContext context) {
        super(response, context);
    }

    public BrowseCouponsVerification hasCouponWithCode(String couponCodeAlias) {
        findCouponByCode(couponCodeAlias);
        return this;
    }

    public BrowseCouponsVerification couponHasDiscountRate(String couponCodeAlias, double expectedDiscountRate) {
        var coupon = findCouponByCode(couponCodeAlias);
        assertThat(coupon.getDiscountRate())
                .as("Discount rate for coupon '%s'", couponCodeAlias)
                .isEqualTo(expectedDiscountRate);
        return this;
    }

    public BrowseCouponsVerification couponHasValidFrom(String couponCodeAlias, String expectedValidFrom) {
        var coupon = findCouponByCode(couponCodeAlias);
        var expectedInstant = Converter.toInstant(expectedValidFrom);
        assertThat(coupon.getValidFrom())
                .as("ValidFrom for coupon '%s'", couponCodeAlias)
                .isEqualTo(expectedInstant);
        return this;
    }

    public BrowseCouponsVerification couponHasValidTo(String couponCodeAlias, String expectedValidTo) {
        var coupon = findCouponByCode(couponCodeAlias);
        var expectedInstant = Converter.toInstant(expectedValidTo);
        assertThat(coupon.getValidTo())
                .as("ValidTo for coupon '%s'", couponCodeAlias)
                .isEqualTo(expectedInstant);
        return this;
    }

    public BrowseCouponsVerification couponHasUsageLimit(String couponCodeAlias, int expectedUsageLimit) {
        var coupon = findCouponByCode(couponCodeAlias);
        assertThat(coupon.getUsageLimit())
                .as("Usage limit for coupon '%s'", couponCodeAlias)
                .isEqualTo(expectedUsageLimit);
        return this;
    }

    public BrowseCouponsVerification couponHasUsedCount(String couponCode, int expectedUsedCount) {
        var coupon = findCouponByCode(couponCode);
        assertThat(coupon.getUsedCount())
                .as("Used count for coupon '%s'", couponCode)
                .isEqualTo(expectedUsedCount);
        return this;
    }

    private BrowseCouponsResponse.CouponDto findCouponByCode(String couponCodeAlias) {
        assertThat(couponCodeAlias)
                .as("Coupon code alias parameter")
                .isNotNull();

        assertThat(getResponse())
                .as("Response should not be null")
                .isNotNull();
        assertThat(getResponse().getCoupons())
                .as("Coupons list should not be null")
                .isNotNull();

        var couponCode = getContext().getParamValue(couponCodeAlias);

        return getResponse().getCoupons().stream()
                .filter(c -> couponCode.equals(c.getCode()))
                .findFirst()
                .orElseThrow(() -> new AssertionError(
                        String.format("Coupon with code '%s' not found. Available coupons: %s",
                                couponCode,
                                getResponse().getCoupons().stream()
                                        .map(BrowseCouponsResponse.CouponDto::getCode)
                                        .toList())));
    }
}

File: system-test/dotnet/Dsl.Core/UseCase/UseCases/BrowseCouponsVerification.cs
using Dsl.Core.Shared;
using Shouldly;
using Driver.Port.Dtos;
using Common;

namespace Dsl.Core.UseCase.UseCases;

public class BrowseCouponsVerification : ResponseVerification<BrowseCouponsResponse>
{
    public BrowseCouponsVerification(BrowseCouponsResponse response, UseCaseContext context)
        : base(response, context)
    {
    }

    public BrowseCouponsVerification HasCouponWithCode(string couponCodeAlias)
    {
        FindCouponByCode(couponCodeAlias);
        return this;
    }

    public BrowseCouponsVerification CouponHasDiscountRate(string couponCodeAlias, decimal expectedDiscountRate)
    {
        var coupon = FindCouponByCode(couponCodeAlias);
        coupon.DiscountRate.ShouldBe(expectedDiscountRate, $"Expected coupon '{couponCodeAlias}' to have discount rate {expectedDiscountRate:F2}");
        return this;
    }

    public BrowseCouponsVerification CouponHasValidFrom(string couponCodeAlias, string? expectedValidFrom)
    {
        var coupon = FindCouponByCode(couponCodeAlias);
        DateTime? expectedValidFromDateTime = Converter.ToDateTime(expectedValidFrom);
        coupon.ValidFrom.ShouldBe(expectedValidFromDateTime, $"Expected coupon '{couponCodeAlias}' to have validFrom '{expectedValidFrom}'");
        return this;
    }

    public BrowseCouponsVerification CouponHasValidTo(string couponCodeAlias, string? expectedValidTo)
    {
        var coupon = FindCouponByCode(couponCodeAlias);
        DateTime? expectedValidToDateTime = Converter.ToDateTime(expectedValidTo);
        coupon.ValidTo.ShouldBe(expectedValidToDateTime, $"Expected coupon '{couponCodeAlias}' to have validTo '{expectedValidTo}'");
        return this;
    }

    public BrowseCouponsVerification CouponHasUsageLimit(string couponCodeAlias, int expectedUsageLimit)
    {
        var coupon = FindCouponByCode(couponCodeAlias);
        coupon.UsageLimit.ShouldBe(expectedUsageLimit, $"Expected coupon '{couponCodeAlias}' to have usage limit {expectedUsageLimit}");
        return this;
    }

    public BrowseCouponsVerification CouponHasUsedCount(string couponCode, int expectedUsedCount)
    {
        var coupon = FindCouponByCode(couponCode);
        coupon.UsedCount.ShouldBe(expectedUsedCount, $"Expected coupon '{couponCode}' to have used count {expectedUsedCount}");
        return this;
    }

    private CouponDto FindCouponByCode(string couponCodeAlias)
    {
        Response.ShouldNotBeNull();
        Response.Coupons.ShouldNotBeNull("No coupons found in response");

        var couponCode = Context.GetParamValue(couponCodeAlias);

        var coupon = Response.Coupons.FirstOrDefault(c => string.Equals(couponCode, c.Code));

        coupon.ShouldNotBeNull($"Coupon with code '{couponCode}' not found. Available coupons: [{string.Join(", ", Response.Coupons.Select(c => c.Code))}]");

        return coupon;
    }
}
